component/kafka/simple/component.go

```diff
@@ -0,0 +1,361 @@
+	batchTimeout             time.Duration
+	retries                  uint
+	retryWait                time.Duration
+	partitionConsumerFactory func(ctx context.Context) (sarama.Client, sarama.Consumer, map[int32]sarama.PartitionConsumer, error)
+	// DurationOffset options
+	durationBasedConsumer bool
+	durationOffset        time.Duration
+	timeExtractor         func(*sarama.ConsumerMessage) (time.Time, error)
+	// NotificationOnceReachingLatestOffset options
+	latestOffsetReachedChan chan<- struct{}
+	startingOffsets         map[int32]int64
+	latestOffsets           map[int32]int64
+	once                    sync.Once
+}
+
+// New initializes a new simple kafka consumer component with support for functional configuration.
+func New(name string, brokers []string, topic string, proc kafka.BatchProcessorFunc, saramaCfg *sarama.Config, oo ...OptionFunc) (*Component, error) {
+	var errs []error
+	if name == "" {
+		errs = append(errs, errors.New("name is required"))
+	}
+
+	if validation.IsStringSliceEmpty(brokers) {
+		errs = append(errs, errors.New("brokers are empty or have an empty value"))
+	}
+
+	if topic == "" {
+		errs = append(errs, errors.New("topic is required"))
+	}
+
+	if proc == nil {
+		errs = append(errs, errors.New("work processor is required"))
+	}
+
+	if len(errs) > 0 {
+		return nil, patronErrors.Aggregate(errs...)
+	}
+
+	cmp := &Component{
+		name:         name,
+		brokers:      brokers,
+		topic:        topic,
+		proc:         proc,
+		retries:      defaultRetries,
+		retryWait:    defaultRetryWait,
+		batchSize:    defaultBatchSize,
+		batchTimeout: defaultBatchTimeout,
+		failStrategy: defaultFailureStrategy,
+		saramaConfig: saramaCfg,
+	}
+
+	cmp.partitionConsumerFactory = cmp.createPartitionConsumers
+
+	for _, optionFunc := range oo {
+		err := optionFunc(cmp)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return cmp, nil
+}
+
+// Run starts the consumer processing loop to process messages from Kafka.
+func (c *Component) Run(ctx context.Context) error {
+	retries := int(c.retries)
+
+	if retries == 0 {
+		return c.runWithoutRetry(ctx)
+	}
+	return c.runWithRetry(ctx, retries)
+}
+
+func (c *Component) runWithoutRetry(ctx context.Context) error {
+	_, err := c.processing(ctx)
+	if err != nil {
+		return fmt.Errorf("simple consumer error on topic %s: %w", c.topic, err)
+	}
+	return nil
+}
+
+func (c *Component) runWithRetry(ctx context.Context, retries int) error {
+	var hasProcessedMessages bool
+	var err error
+
+	for i := 0; ; i++ {
+		hasProcessedMessages, err = c.processing(ctx)
+		if err != nil {
+			log.Errorf("simple consumer error on topic %s: %v", c.topic, err)
+		}
+
+		if isContextDone(ctx) {
+			break
+		}
+		if hasProcessedMessages {
+			i = 0
+		}
+		if i >= retries {
+			break
+		}
+
+		log.Errorf("failed run, retry %d/%d with %v wait", i, c.retries, c.retryWait)
+		time.Sleep(c.retryWait)
+	}
+
+	return err
+}
+
+func isContextDone(ctx context.Context) bool {
+	select {
+	case _, ok := <-ctx.Done():
```

Comment:
Is this a typo? I think you need only this:
```suggestion
	case <-ctx.Done():
```
(and no `!ok` check). Quoting documentation:
```
	// Done returns a channel that's closed when work done on behalf of this
	// context should be canceled. Done may return nil if this context can
	// never be canceled. Successive calls to Done return the same value.
	// The close of the Done channel may happen asynchronously,
	// after the cancel function returns.
```

Successive calls to Done return the same value." -> so it does not matter whether it is closed or not, either we are block-waiting on the channel or channel returns a value/is closed (idempotent conditions).

However, I see two additional issues here:
1. this function is hiding that a channel is being polled, which is not very Go-idiomatic
2. no test caught this bug

Can you think of a way to not hide a select-with-default within a function and perhaps use it in other `select`-like statements on caller side?

Reply:
Makes sense.

> this function is hiding that a channel is being polled, which is not very Go-idiomatic

I may tend to disagree with the idiomatic opinion (most discussions around what idiomatic means tend to lead to personal opinions ^^). Here's an example in [k8s](https://github.com/kubernetes/kubernetes/blob/70f745a37aba9cdc6a83c17600f5c9c87425c941/cmd/kube-scheduler/app/server.go#L155-L164).

> no test caught this bug

Which bug? The behavior was [ok](https://go.dev/play/p/-R7sOE7SG7O) nonetheless I think.

Reply:
Sure; my point is that it hides (channels) control flow, I thought it is not a good Go practice to do that.

> no test caught this bug
> 
> Which bug? The behavior was [ok](https://go.dev/play/p/-R7sOE7SG7O) nonetheless I think.

Cool; then I was wrong, I thought it would return differently.